### PR TITLE
Vol surface implementation 

### PR DESCRIPTION
**Pull request by sefhamada**
Fixes #.
 
- Added  "VolSurface" class and an interpolation method (with its generic) for term/strike type of vol.
- The main changes are in the file : vol_surface_class.R
- I added documentations and examples similar to the existing classes in fmbasics
- I added the test harness in testthat folder: test_volsurface.R
- I  performed devtool::check() with 0 error, 0 warnings and 0 notes.

@imanuelcostigan

### CODE REVIEW

File: R/utils.R
Comment (imanuelcostigan) [inline R/utils.R]: Probably something to remove?

File: R/generic-methods.R
Comment (imanuelcostigan) [inline R/generic-methods.R]: Why can't you extend `interpolate()` by implementing a specific method for `VolSurface`?

File: R/utils.R
Comment (imanuelcostigan) [inline R/utils.R]: Why do you need to hard code this - isn't this in `volsurface.csv"?

File: R/utils.R
Comment (imanuelcostigan) [inline R/utils.R]: You don't need to pass the result to `vol_obj` and then return it on the next line. Leave the call to `VolSurface()` as the last line in the function.

File: R/vol-surface-class.R
Comment (imanuelcostigan) [inline R/vol-surface-class.R]: you have one too many back-ticks here.

File: R/vol-surface-class.R
Comment (imanuelcostigan) [inline R/vol-surface-class.R]: spelling mistake here.

File: R/vol-surface-class.R
Comment (imanuelcostigan) [inline R/vol-surface-class.R]: we shouldnt be using matrices as an input type - they are terrible to work with. Use a (tidy) tibble. Suggest the tibble structure is something like:

```csv
maturity, strike, value
20200201, 72.00, 0.20
20200301, 72.00, 0.24
..., ..., ...
20200201, 82.00, 0.25
20200301, 82.00, 0.26
..., ..., ...
```

In the doc be very clear about the structure of this including for example that:

1. The name of the second column (e.g. `strike` or `moneyness` or `delta`) will define what sort of surface you are getting. Be clear about the options that are valid.
2. The string values / formats required for each field: e.g. `maturity` must be a string of format `YYYYMMDD`, `value` much be in decimals...etc.

File: R/vol-surface-class.R
Comment (imanuelcostigan) [inline R/vol-surface-class.R]: remove comments like this.

File: R/vol-surface-class.R
Comment (imanuelcostigan) [inline R/vol-surface-class.R]: Once you are much clearer on what `vol_quotes` should be, you will be able to simplify this interface considerably.

File: R/vol-surface-class.R
Comment (imanuelcostigan) [inline R/vol-surface-class.R]: why are you including an assertion in a `new_X()` method? Shouldn't this what the `validate_X()` function is there for?

File: R/vol-surface-class.R
Comment (imanuelcostigan) [inline R/vol-surface-class.R]: Once you are clearer on `vol_type` structure per above, you will be able to update these assertions.

File: R/vol-surface-class.R
Comment (imanuelcostigan) [inline R/vol-surface-class.R]: Not required. This is method will only be called when `x` is a `VolSurface`

File: R/vol-surface-class.R
Comment (imanuelcostigan) [inline R/vol-surface-class.R]: Can you please use the same language as we use for `interpolate.ZeroCurve()`? THe interface should look like 

```R
function(x, at, ...)
```

where `at` is a named list of length two where each list member is of the same length as the other and can be a vector of terms and strikes (e.g. `list(term = c(1, 2), smile = c(72, 92))`).

File: R/vol-surface-class.R
Comment (imanuelcostigan) [inline R/vol-surface-class.R]: Why are these comments in here?

File: R/vol-surface-class.R
Comment (imanuelcostigan) [inline R/vol-surface-class.R]: Is this if statement necessary? Why isn't this necessary for the `ZeroCurve` class?

File: R/vol-surface-class.R
Comment (imanuelcostigan) [reply R/vol-surface-class.R]: I think this method needs a bit of work to address other comments above. Once this is done, I will review the rest of this.

File: R/zero-curve-class.R
Comment (imanuelcostigan) [inline R/zero-curve-class.R]: Do you really need a new interpolation method called `LinearTimeVarInterpolation`? Isn't this simply LinearInterpolation applied to time weighted variance?

File: inst/extdata/volsurface.csv
Comment (imanuelcostigan) [inline inst/extdata/volsurface.csv]: You may need to update the format of this data (or write another function to transform this data into the format suggested in one of my earlier comments)

File: inst/extdata/zerocurve.csv
Comment (imanuelcostigan) [inline inst/extdata/zerocurve.csv:1]: why has this file changed?

File: tests/testthat/test-volsurface.R
Comment (imanuelcostigan) [inline tests/testthat/test-volsurface.R]: This needs to be slightly more comprehensive. Take your queue from the `test-interpolate.R` file

**Review by imanuelcostigan — CHANGES_REQUESTED** [review]
@sefhamada 

- [ ] Please format all files using `styler::style_file()`. 
- [ ] Please reflected on my in-line comments and update your PR. 

Thanks.

File: R/utils.R
Comment (sefhamada) [reply R/utils.R]: done

File: R/utils.R
Comment (sefhamada) [reply R/utils.R]: The data in "volsurface.csv" does not include this information (ref date, tenors, strike) explicitly. It only contains the tenor dates in the form yyyymmdd, the strikes are given via moneyness: X5., X10., X15., ...., X250., where the digits refer to (K/S)*100. This is an example of a vol surface extracted from Murex.

File: R/utils.R
Comment (sefhamada) [reply R/utils.R]: That's true, I removed the extra variable assignment. It is a cautious habit.

File: R/vol-surface-class.R
Comment (sefhamada) [reply R/vol-surface-class.R]: done

File: R/vol-surface-class.R
Comment (sefhamada) [reply R/vol-surface-class.R]: fixed dimensions spelling in line 4 and 7.

File: R/vol-surface-class.R
Comment (sefhamada) [reply R/vol-surface-class.R]: I used this "If" statement as the interpolation technique depends on the type of the surface. At this stage we are only accounting for "strike/term" type of surface, and left a place-holder for interpolating other types (delta/term, moneyness/term).

File: R/vol-surface-class.R
Comment (sefhamada) [reply R/vol-surface-class.R]: Here I am using the same logic used in ZeroCurve class where the assertion of the interpolation method is introduced in new_ZeroCurve rather than in validate_ZeroCurve. We can easily shift the assertion clause to the validation function.

File: R/vol-surface-class.R
Comment (sefhamada) [reply R/vol-surface-class.R]: True. Now removed.

File: R/vol-surface-class.R
Comment (sefhamada) [reply R/vol-surface-class.R]: I changed the structure of VolSurface following the above suggestion. Now it only includes: reference_date, vol_quotes(tibble as described in the example above), ticker and Interpolation.

File: R/vol-surface-class.R
Comment (sefhamada) [reply R/vol-surface-class.R]: done

File: R/vol-surface-class.R
Comment (sefhamada) [reply R/vol-surface-class.R]: Now the constructor has 4 members, "reference_date", "vol_quotes", "ticker", "interpolation".

File: R/vol-surface-class.R
Comment (sefhamada) [reply R/vol-surface-class.R]: done

File: R/vol-surface-class.R
Comment (sefhamada) [reply R/vol-surface-class.R]: Please feel free to suggest any improvement to the method. It does look slightly different now that the above comments are incorporated.

File: R/vol-surface-class.R
Comment (sefhamada) [reply R/vol-surface-class.R]: I am aware there are many ways of interpolating Volatilities. I used the "if" statement as the implemented approach is only valid when the interpolation is of type (linear on term, cubic spline on strike) over a time-variance surface, and the surface is of type term/strike.

File: R/zero-curve-class.R
Comment (sefhamada) [reply R/zero-curve-class.R]: Even though the name says Linear, but the actual interpolation is Linear on time, Cubic spline on strike. Including the whole description will blow up the length of the method name.  The actual interpolation is two-fold: 1st interpolation on the time, and a 2nd interpolation on the strikes.

File: inst/extdata/volsurface.csv
Comment (sefhamada) [reply inst/extdata/volsurface.csv]: This data is extracted from a Murex report as an example of vol surface data, it is then massaged in R/utils.R to construct an object following the suggested class structure of VolSurface.

File: inst/extdata/zerocurve.csv
Comment (sefhamada) [reply inst/extdata/zerocurve.csv:1]: Not totally sure, but we had to change the  data in the zero curve to fix the bug in the discount factor interpolation.

File: R/generic-methods.R
Comment (sefhamada) [reply R/generic-methods.R]: Good idea ! It is now inherited by VolSurface class. I previously thought to follow the same logic as interpolate_dfs, interpolate_zeros, interpolate_fwd; but this is actually different.

File: tests/testthat/test-volsurface.R
Comment (sefhamada) [reply tests/testthat/test-volsurface.R]: If I didn't misunderstood the comment, I have added 2 more interpolation points.

File: R/vol-surface-class.R
Comment (sefhamada) [reply R/vol-surface-class.R]: Good point ! now the interpolation for VolSurface class inherits from the generic interpolation method, and it is using "at" as per the suggested example. Although I assumed the names of the list members are "term" and "strike" instead of "smile".

**Comment by sefhamada** [comment]
Just added an interpolation example and print/format for VolSurface objects.

File: R/utils.R
Comment (imanuelcostigan) [reply R/utils.R]: I mean the first line of the `volsurfaces.csv` file has the percentages you have encoded in `strikes`.

File: R/utils.R
Comment (imanuelcostigan) [inline R/utils.R]: You don't need this many new lines at the end of a file.

File: R/vol-surface-class.R
Comment (imanuelcostigan) [reply R/vol-surface-class.R]: The reason for it in ZeroCurve land is because I wanted to make sure when an instance of this was created, it would have a functioning and calibrated interpolation `interpolator`.

File: R/vol-surface-class.R
Comment (imanuelcostigan) [inline R/vol-surface-class.R]: I don't think `vol_quotes` should be defined in such a way that the column order matters.

File: R/vol-surface-class.R
Comment (imanuelcostigan) [inline R/vol-surface-class.R]: If you move this up a few assertions, you can ensure that `colnames()` calls make sense.

File: R/vol-surface-class.R
Comment (imanuelcostigan) [reply R/vol-surface-class.R]: Should you create a new class `VolQuotes` (and possibly an extension `EquityVolQuotes`) that extends `tibble`? That way you can move these assertions out of VolSurface and into VolQuotes (or extension thereof).

File: R/vol-surface-class.R
Comment (imanuelcostigan) [inline R/vol-surface-class.R]: This seems like an anti-pattern. You probably want to ensure that date fields in vol_quotes are actually dates or date-times. Leave it to the user to make sure their inputs are correct rather than giving them the flexibility (and hence introducing greater surface for bugs to crawl out of).

File: R/vol-surface-class.R
Comment (imanuelcostigan) [inline R/vol-surface-class.R]: Why not `is.numeric()`?

File: R/vol-surface-class.R
Comment (imanuelcostigan) [inline R/vol-surface-class.R]: This is another example of where your anti-pattern above reemerges.

File: R/vol-surface-class.R
Comment (imanuelcostigan) [inline R/vol-surface-class.R]: Use `FALSE` instead of `F` and `TRUE` instead of `T`.

File: R/vol-surface-class.R
Comment (imanuelcostigan) [inline R/vol-surface-class.R]: You have already checked for this in `new_VolSurface()`

File: R/vol-surface-class.R
Comment (imanuelcostigan) [inline R/vol-surface-class.R]: No line spacing required here

File: R/vol-surface-class.R
Comment (imanuelcostigan) [inline R/vol-surface-class.R:173]: Or here

File: R/vol-surface-class.R
Comment (imanuelcostigan) [inline R/vol-surface-class.R]: You don't need to qualify the package namespace in examples contained in the same package

File: R/vol-surface-class.R
Comment (imanuelcostigan) [inline R/vol-surface-class.R]: Not necessary because `new_VolSurface()` doesn't permit other options

File: R/vol-surface-class.R
Comment (imanuelcostigan) [inline R/vol-surface-class.R]: Is there a better way of handling situations where quotes are not according to strike?

File: R/vol-surface-class.R
Comment (imanuelcostigan) [inline R/vol-surface-class.R]: This doesn't seem right. the length of your `x` values can be different to the length of your `y` values. What is the behaviour of `stats::approxfun()` in this case?

File: R/vol-surface-class.R
Comment (imanuelcostigan) [inline R/vol-surface-class.R]: I think it may be more efficient (and consistent with `ZeroCurve`) if the calibration interpolator function is stored in the instance of `VolSurface`. Some part of this shouldn't change for a given `(term, strike)` that you are looking to interpolate.

File: R/vol-surface-class.R
Comment (imanuelcostigan) [inline R/vol-surface-class.R]: This seems to be the way you are trying to deal with `vol_quotes` being a tidy tibble. See next comment - is this really right?

File: tests/testthat/test-volsurface.R
Comment (imanuelcostigan) [reply tests/testthat/test-volsurface.R]: yes, but can you also provide some interpolation points that include the region before the first strike and after the last strike / maturity date?

**Review by imanuelcostigan — CHANGES_REQUESTED** [review]
@sefhamada See comments below. There are some further things to correct / change and I still have a few qs. I also think I have overestimated styler's abilities out of the box. There is some code style that isn't quite right. I'll fix this up once we merge this PR into master.

File: inst/extdata/zerocurve.csv
Comment (imanuelcostigan) [reply inst/extdata/zerocurve.csv:1]: You need to explain why this changed (didn't it work previously?).

File: R/utils.R
Comment (sefhamada) [reply R/utils.R]: image

 This is the format of  "vols" data sourced from the exdata (volsurface.csv). I now use the first line (colnames) containing the percentage information to retrieve the moneyness and therefore the strikes.

File: R/vol-surface-class.R
Comment (sefhamada) [reply R/vol-surface-class.R]: Now the assertion on the Interpolation is included in the validation function.

File: inst/extdata/zerocurve.csv
Comment (sefhamada) [reply inst/extdata/zerocurve.csv:1]: What happened here is: The data in zerocurve.csv was wrong (wrong pillar points) which caused the zero interpolation of build_zero_curve() to return wrong results. So we have fixed the bug in the commit  "sefhamada:fixing-the-zero-curve" by changing the data in the file. This was done in the master branch. So I have changed the zerocurve.csv data in this fork to match the data in the master.

File: tests/testthat/test-volsurface.R
Comment (sefhamada) [reply tests/testthat/test-volsurface.R]: Added a point with strike = 3, on the left of the interpolation support; and a point with strike = 300 on the right of the support with maturity = "2025-06-10" beyond the maturities of the surface data.

File: R/utils.R
Comment (sefhamada) [reply R/utils.R]: Blank lines now removed.

File: R/vol-surface-class.R
Comment (sefhamada) [reply R/vol-surface-class.R]: done

File: R/vol-surface-class.R
Comment (sefhamada) [reply R/vol-surface-class.R]: Now all the assertions are moved to validate_VolSurface()

File: R/vol-surface-class.R
Comment (sefhamada) [reply R/vol-surface-class.R]: now removed.

File: R/vol-surface-class.R
Comment (sefhamada) [reply R/vol-surface-class.R:173]: Removed

File: R/vol-surface-class.R
Comment (sefhamada) [reply R/vol-surface-class.R]: Very true, now removed.

File: R/vol-surface-class.R
Comment (sefhamada) [reply R/vol-surface-class.R]: I would have used numeric, I am using the suggestions from the comment made above being decimal:
image

Comment (sefhamada) [reply R/vol-surface-class.R]: We want the vols to be in decimals and not in percentage, so it is just using numeric.

File: R/vol-surface-class.R
Comment (sefhamada) [reply R/vol-surface-class.R]: Now created a VolQuotes class that contains the volatility information, for Equity there is no need to extend, however, this class could be inherited for other markets - e.g. interest rates where the tenors dimension is also needed.

File: R/vol-surface-class.R
Comment (sefhamada) [reply R/vol-surface-class.R]: Now these assertions are moved to VolQuotes object validation; and the user has to provide the data with sorted input dimensions similar to ZeroCurve, otherwise it would be messy.

File: R/vol-surface-class.R
Comment (sefhamada) [reply R/vol-surface-class.R]: The class of VolQuotes is now designed to to accommodate different types of smiles (strike, delta, moneyness). The user can specify the type as a controlled input.

File: R/vol-surface-class.R
Comment (sefhamada) [reply R/vol-surface-class.R]: now the class VolQuotes insures that the maturities are of class date.

File: R/vol-surface-class.R
Comment (sefhamada) [reply R/vol-surface-class.R]: The logic of interpolation is now changed so that only accept the pre-implemented methods.

File: R/vol-surface-class.R
Comment (sefhamada) [reply R/vol-surface-class.R]: This now is the case where the interpolator is embedded in the VolSurface object - calling a separate mathematical Linear_Cubic implemented in utils.R file. Similar to ZeroCurve class logic.

File: R/vol-surface-class.R
Comment (sefhamada) [reply R/vol-surface-class.R]: The main conditions for the vols data are sorted dimensions and for the data to be a rectangular data.

**Review by imanuelcostigan — APPROVED** [review]
I will merge this PR into a branch on `fmbasics` and do a bit more work on it before I merge into master. I'll tag you when I'm done.

**Pull request merged**
